File: src/remove_duplicate_ut_entries.rb
#! /usr/bin/env ruby
# coding: utf-8

# Author: UTUMI Hirosi (utuhiro78 at yahoo dot co dot jp)
# License: Apache License, Version 2.0

require 'open-uri'
require 'nkf'

targetfiles = ARGV

if ARGV == []
	puts "Usage: ruby script.rb [FILE]"
	exit
end

url = "https://raw.githubusercontent.com/google/mozc/master/src/data/dictionary_oss/id.def"
id_mozc = URI.open(url).read.split(" 名詞,一般,")[0]
id_mozc = id_mozc.split("\n")[-1]

filename = ARGV[0]
dicname = filename

file = File.new(filename, "r")
	lines = file.read.split("\n")
file.close

lines.length.times do |i|
	s1 = lines[i].split("	")

	# IDを最新のものに変更
	s1[1..2] = [id_mozc, id_mozc]

	# 読みと表記を先頭にする
	s1 = [s1[0], s1[4], s1[3], s1[1],s1[2]]
	lines[i] = s1.join("	")
end

lines = lines.sort

dicfile = File.new(dicname, "w")

lines.length.times do |i|
	s1 = lines[i].split("	")
	s2 = lines[i - 1].split("	")

	# UT辞書内で重複するエントリのうち、コスト値の大きいものをスキップ
	# あいおい	相生	1847	1847	8200
	# あいおい	相生	1847	1847	8400
	if s1[0..1] == s2[0..1]
		next
	end

	s1 = [s1[0], s1[3], s1[4], s1[2], s1[1]]
	dicfile.puts s1.join("	")
end

dicfile.close

File: src/remove_duplicate_ut_entries.py
#!/usr/bin/env python
# coding: utf-8

# Author: UTUMI Hirosi (utuhiro78 at yahoo dot co dot jp)
# License: Apache License, Version 2.0

import urllib.request
import sys

if len(sys.argv) == 1:
	print("Usage: python script.py [FILE]")
	exit()

file_name = sys.argv[1]

# Mozc の一般名詞のID
url = "https://raw.githubusercontent.com/google/mozc/master/src/data/dictionary_oss/id.def"
with urllib.request.urlopen(url) as response:
	id_mozc = response.read().decode()

id_mozc = id_mozc.split(" 名詞,一般,")[0].split("\n")[-1]

with open(file_name, "r", encoding="utf-8") as file:
	lines = file.read().splitlines()

for i in range(len(lines)):
	entry = lines[i].split("\t")

	# IDを最新のものに変更
	entry[1:3] = [id_mozc, id_mozc]

	# 読みと表記を先頭にする
	entry.insert(1, entry[4])
	lines[i] = entry[:5]

lines.sort()

with open(file_name, "w", encoding="utf-8") as dict_file:
	for i in range(len(lines)):
		# [読み, 表記] が重複するエントリのうち、コストの大きいものをスキップ
		# あいおい	相生	1843	1843	8200
		# あいおい	相生	1843	1843	8400
		if lines[i][0:2] == lines[i - 1][0:2]:
			continue

		entry = lines[i].copy()
		entry.append(entry[1])
		entry.pop(1)
		dict_file.write("\t".join(entry) + "\n")
